### PR TITLE
Make `Pnpm` separate from `Npm`

### PR DESCRIPTION
**Pull request by fviernau**
Rewrite large parts of Pnpm from scratch to figure out all information based on `pnpm` CLI output, and
remove the inheritance from `Npm`. 

Part of #9261.

### CODE REVIEW

File: plugins/package-managers/node/src/main/kotlin/Npm.kt [hunk 1]
```diff
@@ -241,7 +241,7 @@ open class Npm(
         }
     }
 
-    internal open fun getRemotePackageDetails(workingDir: File, packageName: String): PackageJson? {
+    protected open fun getRemotePackageDetails(workingDir: File, packageName: String): PackageJson? {
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/Npm.kt]: The commit message does not seem to match the diff. Also, typo in "unsucessful".

File: plugins/package-managers/node/src/funTest/kotlin/PnpmFunTest.kt [hunk 2]
```diff
@@ -26,6 +26,7 @@ import org.ossreviewtoolkit.analyzer.analyze
 import org.ossreviewtoolkit.analyzer.create
 import org.ossreviewtoolkit.analyzer.resolveSingleProject
 import org.ossreviewtoolkit.model.toYaml
+import org.ossreviewtoolkit.plugins.packagemanagers.node.pnpm.Pnpm
```
Comment (sschuberth) [inline plugins/package-managers/node/src/funTest/kotlin/PnpmFunTest.kt:29]: This should move the test to a sub-package as well.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/project-with-lockfile-expected-output.yml [hunk 3]
```diff
@@ -47,7 +47,7 @@ project:
         - id: "NPM::domhandler:2.4.2"
           dependencies:
           - id: "NPM::domelementtype:1.3.1"
-        - id: "NPM::domutils:1.5.1"
+        - id: "NPM::domutils:1.7.0"
```
Comment (sschuberth) [inline plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/project-with-lockfile-expected-output.yml:50]: Commit message: "that it *is* good to drop" (?)

Comment (sschuberth) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/project-with-lockfile-expected-output.yml:50]: "Omiting" -> "Omitting"

File: plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                getRemotePackageDetails = pnpm::getRemotePackageDetails
+            )
+        }
+
+    private fun readPackageJson(packageJsonFile: File): PackageJson =
+        packageJsonCache.getOrPut(packageJsonFile.realFile()) { parsePackageJson(packageJsonFile) }
+}
+
+private val Dependency.workingDir: File get() = File(path)
+private val Dependency.packageJsonFile: File get() = File(path, "package.json")
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt:81]: How about instead adding this to the class body of `Dependency`:

        @Transient
        val workingDir = File(path)

        @Transient
        val packageJsonFile = workingDir.resolve("package.json")

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt (hunk 4 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt]: I would like to separate upstream data model from any ORT properties for clarify. So, I prefer to keep these extensions.

File: plugins/package-managers/node/src/funTest/kotlin/PnpmFunTest.kt (hunk 2 above)
Comment (fviernau) [reply plugins/package-managers/node/src/funTest/kotlin/PnpmFunTest.kt:29]: Shall we really do this? We'll end up with one file per package.

File: plugins/package-managers/node/src/main/kotlin/Npm.kt (hunk 1 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/Npm.kt]: Dropped the accidental commit.

File: plugins/package-managers/node/src/funTest/kotlin/PnpmFunTest.kt (hunk 2 above)
Comment (sschuberth) [reply plugins/package-managers/node/src/funTest/kotlin/PnpmFunTest.kt:29]: So far we have always aligned the package of the test with the package of the class under test. So yes, I think we should continue doing this, even if it results in having a single test in a package.

**Comment by sschuberth** [comment]
I guess this will get rebased onto https://github.com/oss-review-toolkit/ort/pull/9341 once that is merged to make some commits go away.

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/project-with-lockfile-expected-output.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/project-with-lockfile-expected-output.yml:50]: > The cyclic reference from the workspace root project to itself is now included.

Maybe a bit off-topic, why why is there such a cyclic dependency to begin with? It sounds somewhat wrong to have it at all.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt [hunk 5]
```diff
@@ -52,35 +62,116 @@ class Pnpm(
...
-            "--shamefully-hoist", // Build a similar node_modules structure as NPM and Yarn does.
             workingDir = workingDir
         )
 
     override fun beforeResolution(definitionFiles: List<File>) =
         // We do not actually depend on any features specific to a PNPM version, but we still want to stick to a
         // fixed major version to be sure to get consistent results.
         checkVersion()
+
+    internal fun getRemotePackageDetails(workingDir: File, packageName: String): PackageJson? {
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:147]: Couldn't this be `private` instead of `internal`?

File: plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/project-with-lockfile-expected-output.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply plugins/package-managers/node/src/funTest/assets/projects/synthetic/pnpm/project-with-lockfile-expected-output.yml:50]: I investigated this a bit by removing this circular dependency from the `package.json`.
As a result the `old` implementation did not find any dependencies. Maybe this was a workaround to
make the old `pnpm` implementation "work" with the workspace. If this would really be the case, then it would
mean that nobody is using the pnpm workspaces feature, e.g. would have expected a bug report.

I deliberately didn't remove the cyclic dependency in this PR, as I wanted to make this transparent.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt (hunk 5 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:147]: No, because `NpmDeendencyHandler` is making use of it.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt [hunk 6]
```diff
@@ -52,35 +62,116 @@ class Pnpm(
...
+            parsePackageJson(process.stdout)
+        }.onFailure { e ->
+            logger.warn { "Error getting details for $packageName in directory $workingDir: ${e.message.orEmpty()}" }
+        }.getOrNull()?.also {
+            packageDetailsCache[packageName] = it
+        }
+    }
 }
+
+private enum class Scope(val descriptor: String) {
```
[the remark this answers is not in the document]
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:162]: Somehow I like it this way a tad more, so messing with CLI options is limited to this one small function / single small place.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt [hunk 7]
```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 2019 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:2]: On a related note, I only skimmed other the code changes in this file due to a lack of time. As also the expected output changes, it's a bit hard to understand just from the diff whether the implementation is correct. However, I trust that you did the mentioned comparison to NPM results to get confidence that the new expected results are more correct than the old ones.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        val workspaceModuleDirs = getWorkspaceModuleDirs(workingDir)
+        handler.setWorkspaceModuleDirs(workspaceModuleDirs)
+
+        val moduleInfosForScope = Scope.entries.associateWith { scope -> listModules(workingDir, scope) }
+
+        return workspaceModuleDirs.map { projectDir ->
+            val project = parseProject(projectDir.resolve("package.json"), analysisRoot, managerName)
+
+            val scopeNames = Scope.entries.mapTo(mutableSetOf()) { scope ->
+                val scopeName = scope.toString()
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:87]: Wouldn't it be more speaking here to use `descriptor` and avoid the `toString()` override in `Scope` altogether?

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+                val qualifiedScopeName = DependencyGraph.qualifyScope(project, scope.toString())
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:88]: Reuse `scopeName` instead of `scope.toString()`.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        return parsePnpmList(json).mapTo(mutableSetOf()) { File(it.path) }
+    }
+
+    private fun listModules(workingDir: File, scope: Scope): List<ModuleInfo> {
+        val scopeOption = when (scope) {
+            Scope.DEPENDENCIES -> "--prod"
+            Scope.DEV_DEPENDENCIES -> "--dev"
+        }
+
+        val json = run(workingDir, "list", "--json", "--recursive", "--depth", "10000", scopeOption).stdout
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: How can we confidently limit the depth to 10000 now if previously it was -1?

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        packageDetailsCache[packageName]?.let { return it }
+
+        return runCatching {
+            val process = run(workingDir, "info", "--json", packageName)
+
+            parsePackageJson(process.stdout)
+        }.onFailure { e ->
+            logger.warn { "Error getting details for $packageName in directory $workingDir: ${e.message.orEmpty()}" }
+        }.getOrNull()?.also {
+            packageDetailsCache[packageName] = it
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:157]: Again move this to `onSuccess`?

File: plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import org.ossreviewtoolkit.plugins.packagemanagers.node.parsePackageJson
+import org.ossreviewtoolkit.plugins.packagemanagers.node.pnpm.ModuleInfo.Dependency
+import org.ossreviewtoolkit.utils.common.realFile
+
+internal class PnpmDependencyHandler(private val pnpm: Pnpm) : DependencyHandler<Dependency> {
+    private val workspaceModuleDirs = mutableSetOf<File>()
+    private val packageJsonCache = mutableMapOf<File, PackageJson>()
+
+    private fun Dependency.isProject(): Boolean =
+        isInstalled && workingDir.realFile().absoluteFile in workspaceModuleDirs
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt:40]: Drop `absoluteFile` as `realFile()` already returns an absolute path.

File: plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            readPackageJson(dependency.packageJsonFile).version.orEmpty()
+        } else {
+            dependency.version.takeUnless { it.startsWith("link:") || it.startsWith("file:") }.orEmpty()
+        }
+
+        return Identifier(type, namespace, name, version)
+    }
+
+    override fun dependenciesFor(dependency: Dependency): List<Dependency> =
+        (dependency.dependencies + dependency.optionalDependencies).values.filter { it.isInstalled }.toList()
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt:63]: Drop `toList()`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt (hunk 10 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: 10000 prints out more than -1

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt (hunk 7 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:2]: pnpm list outputs a tree. Maybe this helpds with checking. Comparing with npm you can easily reproduce e.g. with a difftool comparing the two expected results.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt (hunk 10 above)
Comment (sschuberth) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: Really? Isn't -1 an alias for "unlimited"?

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: It seem like.

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt (hunk 7 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:2]: Above mentioned comparison should give a good idea that the change is highly likely correct (without verifying the change manually against the list output)

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt (hunk 10 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: Did you try out -1? (I wonder why you are so sure that this works)

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: I tried, maybe I made a mistake...

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: With "it seem like" i responded to "Really?"

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]:
```
pnpm list --help | grep depth
      --depth -1           Display only projects. Useful in a monorepo. `pnpm ls
                           -r --depth -1` lists all projects in a monorepo
      --depth <number>     Max display depth of the dependency tree
      --depth 0            Display only direct dependencies
```

Comment (sschuberth) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: > I wonder why you are so sure that this works

Because it's being used in the existing code. But now I realize the use is in a different context that does not apply here.

Anyway, the question remains why it is valid to artificially limit the depth of the dependency tree now at an (arbitrary?) number of 10000. I would have expected no limit at all, or if there was a limit, it should be configurable via a package manager specific option to go into the direction of https://github.com/oss-review-toolkit/ort/issues/5626.

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: To my understanding there is no way to make the command work without a limit. 
So, using a large value in my opinion makes sense. I don't think 10k is too small.

If you don't like it still, please make a proposal.

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: > Because it's being used in the existing code. But now I realize the use is in a different context that does not apply here.

I don't know what context you are talking about. But previously existing code didn't call this command at all.

Comment (sschuberth) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: > I don't know what context you are talking about. But previously existing code didn't call this command at all.

I was talking about this line:

https://github.com/oss-review-toolkit/ort/blob/507ee30cea999cb6bfd2cdcc2f5bc8e5e4350dd4/plugins/package-managers/node/src/main/kotlin/Pnpm.kt#L59

It *does* use the `--depth` argument for `pnpm list`, but with -1, and in the context of getting workspaces.

Comment (sschuberth) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: > If you don't like it still, please make a counter proposal.

I *already* made two counter proposals: 1. just omit `, "--depth", "10000"` completely, or 2. make the depth value configurable.

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: > I already made two counter proposals:

ok, let me try to explain again why I prefer to not go with these:

1. Does not work because `-1` lists only projects, but not dependencies
2. Making this a) extends the scope of this PR and b) I doubt it is needed, 
    due to the rationale I already provided above, e.g:

> I don't think 10k is too small. In fact it is so large that it is only a theroretical problem IMO. ORT's dependency graph breaks the cycles anyway and I doubt the depth limit could be exceeded without printing cycles.

I still believe this to be a theoretical problem here and 10k works in practically any case. 
Do you fear if some dependencies would still be missing in the output? If this is the case, then let's agree on a higher hard-coded value. But please let's not make this configurable when we can easily avoid it / there is no need.

Comment (sschuberth) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: > 1. Does not work because `-1` lists only projects, but not dependencies

I don't get why this is relevant. I was not proposing to use `-1`, nor does omitting `--depth 10000` completely imply a value of `-1` AFAIU (because as the help states, `-1` does *not* mean "unlimited" as I thought before, but it's a special value to "Display only projects").

So can you please explain why omitting `--depth 10000` would not work?

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: > So can you please explain why omitting `--depth 10000` would not work?

Because then it uses the default which is `-1`.  (Which is why it is relevant, e.g.)

> I don't get why this is relevant.

Comment (sschuberth) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt:120]: Very weird that `pnpm list --help` says "All dependencies are printed by default", but the [docs](https://pnpm.io/cli/list#--depth-number) say "`pnpm ls --depth 0` (default) will list direct dependencies only".

However, apparently we can use "Infinity" as a depth value instead, which worked in my tests. Please use that value as it's less arbitrary and documents the intention.

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt]: `Infinity` is of course much nicer than `10000`. How did you find that one?

Comment (sschuberth) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt]: It's in the linked docs.

**Review by sschuberth — APPROVED** [review]

File: plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt (hunk 10 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/pnpm/Pnpm.kt]: argh
